### PR TITLE
Add indexProvider

### PR DESCRIPTION
**Pull request by RunxinXu**
This PR adds indexProvider, which enables to search artifact by groupId and/or artifactId when auto completing dependency.

In detail, three things have been done:
- Add API in "artifactSearcher" in Java side to support searching artifact by groupId and/or artifactId.
- Add "indexProvider" in ts side
-  Integrate the results returned by centralProvider, localProvider, and indexProvider.

### CODE REVIEW

File: jdtls.ext/com.microsoft.java.maven.plugin/plugin.xml [hunk 1]
```diff
@@ -4,7 +4,8 @@
   <extension point="org.eclipse.jdt.ls.core.delegateCommandHandler">
     <delegateCommandHandler class="com.microsoft.java.maven.handler.DelegateCommandHandler">
       <command id="java.maven.initializeSearcher"/>
-      <command id="java.maven.searchArtifact"/>
+      <command id="java.maven.searchArtifactByClassName"/>
```
Comment (Eskibear) [inline jdtls.ext/com.microsoft.java.maven.plugin/plugin.xml]: consider to use a single entry `java.maven.searchArtifact`. 
Define the structure of the args, e.g. 
```ts
SearchArtifactArgument {
  searchType: enum;
  classname?: string;
  groupId?: string;
  artifactId?: string;
  ...
}
```

File: jdtls.ext/com.microsoft.java.maven.plugin/src/main/java/com/microsoft/java/maven/ArtifactSearcher.java [hunk 2]
```diff
@@ -92,13 +86,36 @@ public static void initialize(String path){
         }
     }
 
+    public static List<ArtifactResult> searchByGA(String groupId, String artifactId, IProgressMonitor monitor) {
```
Comment (Eskibear) [inline jdtls.ext/com.microsoft.java.maven.plugin/src/main/java/com/microsoft/java/maven/ArtifactSearcher.java]: `GA` is not human-readable. consider `Identifier`

File: src/completion/indexProvider.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+import * as vscode from "vscode";
+import { IArtifactSearchResult } from "../artifactSearcher";
+import { executeJavaLanguageServerCommand } from "../jdtls/commands";
+import { COMMAND_COMPLETION_ITEM_SELECTED, INFO_COMPLETION_ITEM_SELECTED } from "./constants";
+import { IMavenCompletionItemProvider } from "./IArtifactProvider";
+import { getSortText } from "./versionUtils";
+
+class IndexProvider implements IMavenCompletionItemProvider {
+    public async getGroupIdCandidates(groupIdHint: string, artifactIdHint: string): Promise<vscode.CompletionItem[]> {
+        const docs: IArtifactSearchResult[] = await executeJavaLanguageServerCommand("java.maven.searchArtifactByGA",
```
Comment (Eskibear) [inline src/completion/indexProvider.ts]: how do you handle `JavaExtensionNotActivatedError` if redhat.java is not present?

File: src/completion/completionProvider.ts [hunk 4]
```diff
@@ -44,8 +45,9 @@ class CompletionProvider implements vscode.CompletionItemProvider {
                 const artifactIdHint: string = artifactIdNode && artifactIdNode.text ? artifactIdNode.text : "";
 
                 const centralItems: vscode.CompletionItem[] = await centralProvider.getGroupIdCandidates(groupIdHint, artifactIdHint);
+                const indexItems: vscode.CompletionItem[] = await indexProvider.getGroupIdCandidates(groupIdHint, artifactIdHint);
```
Comment (Eskibear) [inline src/completion/completionProvider.ts]: Do you need a global indicator like `isJavaExtensionInstalled: boolean`, and calculate the items only when it's true?

File: src/completion/completionProvider.ts [hunk 5]
```diff
@@ -44,8 +45,9 @@ class CompletionProvider implements vscode.CompletionItemProvider {
                 const artifactIdHint: string = artifactIdNode && artifactIdNode.text ? artifactIdNode.text : "";
 
                 const centralItems: vscode.CompletionItem[] = await centralProvider.getGroupIdCandidates(groupIdHint, artifactIdHint);
+                const indexItems: vscode.CompletionItem[] = await indexProvider.getGroupIdCandidates(groupIdHint, artifactIdHint);
                 const localItems: vscode.CompletionItem[] = await localProvider.getGroupIdCandidates(groupIdHint, artifactIdHint);
-                const mergedItems: vscode.CompletionItem[] = this.deDuplicate(centralItems, localItems);
+                const mergedItems: vscode.CompletionItem[] = this.deDuplicate(this.deDuplicate(centralItems, indexItems), localItems);
```
Comment (Eskibear) [inline src/completion/completionProvider.ts]: consider change the signature of `deDuplicate(a, b)` to `deDuplicate(...arr)`?

Comment (RunxinXu) [reply src/completion/completionProvider.ts]: I finds `_.unionBy` api supports union three arrays at the same time. Is it necessary to change it?

File: src/completion/completionProvider.ts (hunk 4 above)
Comment (RunxinXu) [reply src/completion/completionProvider.ts]: ok, I will add it.

File: src/completion/indexProvider.ts (hunk 3 above)
Comment (RunxinXu) [reply src/completion/indexProvider.ts]: By use of global indicator as we have discussed.

File: jdtls.ext/com.microsoft.java.maven.plugin/plugin.xml (hunk 1 above)
Comment (RunxinXu) [reply jdtls.ext/com.microsoft.java.maven.plugin/plugin.xml]: ok.

File: jdtls.ext/com.microsoft.java.maven.plugin/src/main/java/com/microsoft/java/maven/ArtifactSearcher.java (hunk 2 above)
Comment (RunxinXu) [reply jdtls.ext/com.microsoft.java.maven.plugin/src/main/java/com/microsoft/java/maven/ArtifactSearcher.java]: ok.

File: src/completion/completionProvider.ts (hunk 5 above)
Comment (Eskibear) [reply src/completion/completionProvider.ts]: never mind, I'm also ok with current code.

File: src/extension.ts [hunk 6]
```diff
@@ -27,6 +27,7 @@ import { getAiKey, getExtensionId, getExtensionVersion, loadPackageInfo } from "
 import { executeInTerminal } from "./utils/mavenUtils";
 import { openFileIfExists, showTroubleshootingDialog } from "./utils/uiUtils";
 import { Utils } from "./utils/Utils";
+export let isJavaExtensionInstalled: boolean = false;
```
Comment (Eskibear) [inline src/extension.ts]: use https://github.com/microsoft/vscode-maven/blob/b78d316f92b0df4cf73e75b0ed76c056869d00bc/src/jdtls/commands.ts#L19

add a local variable there to cache the value. e.g.

```ts
let isJavaExtensionInstalled: boolean | undefined;
export function isJavaExtEnabled() {
  if (isJavaExtensionInstalled !== undefined) return isJavaExtensionInstalled;
  ...
}
```

File: src/extension.ts [hunk 7]
```diff
@@ -148,6 +149,7 @@ async function doActivate(_operationId: string, context: vscode.ExtensionContext
     const javaExt: vscode.Extension<any> | undefined = extensions.getExtension(EXTENSION_ID);
```
Comment (Eskibear) [inline src/extension.ts]: you can also leverage `isJavaExtEnabled()`.

File: src/artifactSearcher.ts [hunk 8]
```diff
@@ -14,7 +14,8 @@ const UNDIFINED_NAME: string = "570425394";
 const unresolvedCode: string[] = [UNDEFINED_TYPE, UNDIFINED_NAME];
 
 // tslint:disable-next-line: export-name
-export function registerArtifactSearcher(javaExt: vscode.Extension<any>, context: vscode.ExtensionContext): void {
+export function registerArtifactSearcher(context: vscode.ExtensionContext): void {
+    const javaExt: vscode.Extension<any> = <vscode.Extension<any>>getJavaExtension();
```
Comment (Eskibear) [inline src/artifactSearcher.ts]: Potential NPE: javaExt can be undefined.

Comment (Eskibear) [reply src/artifactSearcher.ts]: This exported function should guarantee the correctness by itself.

**Pull request merged**